core/src/main/java/io/grpc/ClientCall.java

```diff
@@ -254,6 +254,20 @@ public void setMessageCompression(boolean enabled) {
     // noop
   }
 
+  /**
+   * Defers flushing the messages written using {@link #sendMessage(Object)} to the
+   * transport until the write buffer is full. The amount of data buffered while corked is
+   * dependent on the transport implementation so users of this API should not make strong
+   * assumptions about how long data is buffered while corked.
+   *
+   * <p>Corking can provide throughput gains by allowing a transport to perform fewer writes to
+   * the underlying network.
+   */
+  @ExperimentalApi
+  public Cork cork() {
```

Comment:
why not just have a method that toggles between corking and uncorking?   i.e. why have object separate from the ClientCall?

Reply:
From my point of view this is far more flexible for the user to implement custom interceptor logic, where you can call `cork` multiple times. Only concern may be allocation of new object.

I also think this API is easier for user to understand. Having both cork / uncork methods may be confusing in terms of what should happen if they're called multiple times (user would have to read javadocs carefully). 

What do you think?

Reply:
On the pro side:
* corking can be stacked
* interceptors don't need to know about other callers of cork

On the con side:
* doesn't match TCP_CORK usages, from which the name is derived
* callers of uncork (`close`) may expect flushes to happen immediately.
* Cork object has different lifetime than Call, and needs to have threading semantics defined


Do you agree with the pros and cons?

Reply:
Agreed on pro side.

Few points for the cons side:
- I think we don't want match TCP_CORK contract, this would involve adding timers inside `Call`. We should leave this to the user to decide it should behave. [1]
- different lifetime: `Cork` is thread safe, but this can indeed result into bad implementation decision and memory leaks if the Cork reference would not be released after the `Call` finishes.

1) `man tcp`
```
If set, don't send out partial frames. All queued partial frames are sent when the option is cleared again. This is useful for prepending headers before calling sendfile(2), or for throughput optimization. As currently implemented, there is a 200 millisecond ceiling on the time for which output is corked by TCP_CORK. If this ceiling is reached, then queued data is automatically transmitted. This option can be combined with TCP_NODELAY only since Linux 2.5.71. This option should not be used in code intended to be portable. 
```

Reply:
@carl-mastrangelo Hi, did you make any decision how the API should look like? I would like to move this forward.